README.md
<!--
SPDX-FileCopyrightText: Night Theme Switcher Contributors
SPDX-License-Identifier: CC-BY-SA-4.0
-->

# Night Theme Switcher <!-- omit in toc -->

![](./res/screenshot.gif)

Automatically toggle your desktop’s color scheme between light and dark, switch backgrounds and run custom commands at sunset and sunrise.

---

## Table of contents <!-- omit in toc -->

- [Graphical installation](#graphical-installation)
- [Command line installation](#command-line-installation)
- [Contributing](#contributing)
- [Frequently asked questions](#frequently-asked-questions)

---

## Graphical installation

Visit [the extension page on extensions.gnome.org](https://extensions.gnome.org/extension/2236/night-theme-switcher/) and enable the extension.

## Command line installation

You will need these tools:

- `meson`
- `gettext`
- `glib-compile-schemas`
- `glib-compile-resources`

Clone the repository and enter the directory:

```bash
git clone https://gitlab.com/rmnvgr/nightthemeswitcher-gnome-shell-extension.git && cd nightthemeswitcher-gnome-shell-extension
```

Build and install using `meson`:

```bash
# System-wide installation
meson setup builddir && meson install -C builddir

# User installation
meson setup builddir --prefix=~/.local && meson install -C builddir
```

Restart your GNOME session and enable the extension:

```bash
gnome-extensions enable nightthemeswitcher@romainvigier.fr
```

## Contributing

You're welcome to contribute to the code or the translations! See [CONTRIBUTING.md](./CONTRIBUTING.md).

## Frequently asked questions

### Some applications don't switch their appearance

Like GNOME's built-in dark mode, the extension switches the standard freedesktop.org color scheme preference. Older applications might not follow this preference.

It is possible to force a GTK theme in the extension preferences, but please note that this can visually break applications. It is better to ask the developers of the applications to support the standard preference.

### Something doesn't work on Ubuntu

Unfortunately, Ubuntu ships heavily modified GNOME components to implement some of its features, like accent colors. Because it does so in a very hacky way instead of working with upstream and other desktop projects on a proper solution, it conflicts with anything that deals with the color scheme and themes, and results in a broken experience with the extension.

As the issue is in Ubuntu, and because I don't have the energy nor the will to work around the consequences of their decisions, unfortunately Ubuntu is not supported until it provides a standard GNOME environment.

### Switching happens too early/too late

There's a hidden setting to configure the offset (in hours) applied to the calculated sunrise and sunset times, you can change it with the `gsettings` command:

```
gsettings --schemadir ~/.local/share/gnome-shell/extensions/nightthemeswitcher@romainvigier.fr/schemas/ set org.gnome.shell.extensions.nightthemeswitcher.time offset $DESIRED_OFFSET
```

### I have disabled Location services but want to use sunrise and sunset times from my location

If you know your coordinates, you can enter them in a hidden setting, and the extension will use them to calculate the sunrise and sunset times. You can set it with the `gsettings` command:

```
gsettings --schemadir ~/.local/share/gnome-shell/extensions/nightthemeswitcher@romainvigier.fr/schemas/ set org.gnome.shell.extensions.nightthemeswitcher.time location '($LATITUDE,$LONGITUDE)'
```


CHANGELOG.md
<!--
SPDX-FileCopyrightText: Night Theme Switcher Contributors
SPDX-License-Identifier: CC-BY-SA-4.0
-->

# Changelog


## [75] - 2023-08-20

This version adds support for GNOME 45 and removes support for GNOME 44.

### Added

- GNOME 45 support

### Removed

- GNOME 44 support

### Changed

- Updated Polish translation by @mbled
- Updated Russian translation by @AHOHNMYC


## [74] - 2023-02-27

This version adds support for GNOME 44 and removes support for GNOME 43.

### Added

- GNOME 44 support

### Removed

- GNOME 43 support


## [73] - 2023-02-27

### Added

- Bulgarian translation by @smarnv
- Estonian translation by @tenseoverflow
- Swedish translation by @bittin1

### Changed

- Updated Arabic translation by Omar TS
- Updated Dutch translation by @philip.goto
- Updated German translation by @seismist
- Updated Spanish translation by @gallegonovato and @oscfdezdz


## [72] - 2022-12-07

### Changed

- Updated Chinese (Simplified) translation by @poipoipoipoipoipoipoipoipoi
- Updated Chinese (Traditional) translation by Sean Wang
- Updated Occitan translation by @quenty_occitania
- Updated Russian translation by @daudix-UFO


## [71] - 2022-11-07

### Changed

- Updated Czech translation by @wolf-wolf
- Updated Italian translation by @starise
- Updated Japanese translation by citrus flavor

### Fixed

- Schedules with sunrise or sunset on a different day are now supported


## [70] - 2022-10-31

### Added

- A notification is sent when the location is unavailable and the manual schedule is used as a fallback

### Changed

- Updated Dutch translation by @Vistaus
- Updated French translation
- Updated German translation
- Updated Turkish translation by @ersen


## [69] - 2022-10-29

### Fixed

- Shell theme was sometimes not changing with the User Themes extension enabled
- Background thumbnail generation was blocking the preferences dialog
- It was possible to set invalid sunset and sunrise times in the preferences dialog


## [68] - 2022-10-28

### Added

- Hidden setting to manually set the location

### Changed

- Updated Czech translation by Vojtěch Perník
- Updated Italian translation by @phaerrax
- Updated Russian translation by @daudix-UFO


## [67] - 2022-09-19

### Changed

- Updated Arabic translation by Omar TS
- Updated Occitan translation by @quenty_occitania


## [66] - 2022-09-08

This version adds support for GNOME 43 and removes support for GNOME 42.

It no longer uses Night Light as a time source, and will instead use the location, if available, to calculate sunrise and sunset times, and will fallback to a fixed schedule if not.

The Shell now having a quick setting to toggle the dark mode made the on-demand time source superfluous, so it has been removed. The keyboard shortcut is now always available.

Note to packagers: Make sure the build type is not `debug` or the extension will print debug logs.

### Added

- Support for GNOME 43
- Czech translation by @wolf-wolf
- Greek translation by Nikolas Tsilivis
- Japanese translation by citrus flavor

### Changed

- Updated Basque translation by @sergitroll9
- Updated Chinese (Traditional) translation by Sean Wang
- Updated Dutch translation by @philip.goto and @Vistaus
- Updated French translation
- Updated Spanish translation by @oscfdezdz

### Removed

- Support for GNOME 42
- Night Light time source
- On-demand time source (replaced by the built-in dark mode quick setting)
- Transition setting


## [65] - 2022-06-17

### Added

- Build option to pack the extension instead of installing it when running `meson install`
- Chinese (Traditional) translation by @olivertzeng

### Changed

- Updated Dutch translation by @philip.goto
- Updated Hungarian translation by @BathoryPeter
- Updated Russian translation by @AHOHNMYC and @daudix-UFO

### Fixed

- Themes under symbolic links were not displayed in the preferences
- Background button did not accept all supported image formats
- Background button did not interpolate thumbnail


## [64] - 2022-04-27

### Added
- Italian translation by @phaerrax

### Changed
- Updated Arabic translation by Omar TS
- Updated Chinese (Simplified) translation by @wsxy162
- Updated Dutch translation
- Updated French translation
- Updated German translation by @sal0max
- Updated Polish translation
- Updated Portuguese (Brazil) translation
- Updated Russian translation by @daudix-UFO
- Updated Spanish translation by @oscfdezdz
- Updated Turkish translation by @ersen

### Fixed
- Background button was not activated when activating a row
- Background button did not support SVG images


## [63] - 2022-04-13

### Changed
- Updated Polish translation by Przemek Juszczyk

### Fixed
- Empty background buttons in the preferences when the monitor size is unavailable


## [62] - 2022-04-08

### Changed
- Updated Chinese (Simplified) translation by @wsxy162
- Updated Dutch translation by @Vistaus
- Updated French translation
- Updated Norwegian Bokmål translation by @kingu
- Updated Turkish translation by @ersen

### Fixed
- On-demand icon not loaded after extension installation
- Adwaita, Adwaita-dark, HighContrast and HighContrastInverse were added to the list of GTK themes even if they were not present on the system
- Chinese (Simplified) translation was not loaded


## [61] - 2022-04-04

### Changed
- Build system has been replaced by Meson to make packaging easier, see the README for the updated installation instructions
- Updated Arabic translation
- Updated Polish translation by Przemek Juszczyk
- Updated Portuguese (Brazil) translation


## [60] - 2022-03-15

### Added
- Polish translation by Tomasz Dróżdż

### Changed
- Updated Chinese (Simplified) translation by poi
- Updated German translation by eladyn


## [59] - 2022-02-26

### Added
- Initial work on Occitan translation by @quenty_occitania


## [58] - 2022-02-25

### Added
- Allow dropping images on background chooser button
- Keep extension enabled on lock screen

### Changed
- Time source settings grouped under an Advanced section

src/modules/Timer.js
// SPDX-FileCopyrightText: 2020-2022 Romain Vigier <contact AT romainvigier.fr>
// SPDX-License-Identifier: GPL-3.0-or-later

import Geoclue from 'gi://Geoclue';
import Gio from 'gi://Gio';
import GLib from 'gi://GLib';
import GObject from 'gi://GObject';
import Meta from 'gi://Meta';
import Shell from 'gi://Shell';

import { gettext as _ } from 'resource:///org/gnome/shell/extensions/extension.js';
import { layoutManager, messageTray, wm } from 'resource:///org/gnome/shell/ui/main.js';
import * as MessageTray from 'resource:///org/gnome/shell/ui/messageTray.js';

import * as debug from '../debug.js';

import { Time } from '../enums/Time.js';


/**
 * The Timer is responsible for signaling any time change to the other modules.
 *
 * They can connect to its 'time' property and query it for the current time.
 *
 * It will try to use the current location as a time source but will fall back
 * to a manual schedule if the location services are disabled or if the user
 * forced the manual schedule in the preferences.
 */
export class Timer extends GObject.Object {
    #settings;
    #interfaceSettings;
    #locationSettings;
    #openPrefs;
    #time;

    #cancellable = null;
    #previousKeybinding = null;
    #timeTimeoutId = null;
    #geoclue = null;
    #geoclueLocationConnectionId = null;
    #suntimesTimeoutId = null;
    #manuallySetTime = false;

    #settingsConnections = [];

    static {
        GObject.registerClass({
            Properties: {
                time: GObject.ParamSpec.string('time', 'Time', 'Time', GObject.ParamFlags.READWRITE, Time.UNKNOWN),
            },
        }, this);
    }

    /**
     * @param {object} params Params object.
     * @param {Gio.Settings} params.settings Timer settings.
     * @param {Function} params.openPrefs Function opening the extension preferences.
     */
    constructor({ settings, openPrefs }) {
        super();
        this.#settings = settings;
        this.#interfaceSettings = new Gio.Settings({ schema: 'org.gnome.desktop.interface' });
        this.#locationSettings = new Gio.Settings({ schema: 'org.gnome.system.location' });
        this.#openPrefs = openPrefs;
    }

    enable() {
        debug.message('Enabling Timer...');
        this.#cancellable = new Gio.Cancellable();
        this.#connectSettings();
        this.#trackTime();
        if (this.#settings.get_boolean('manual-schedule')) {
            debug.message('Using the manual schedule.');
        } else {
            debug.message('Using location.');
            this.#trackLocation();
            this.#trackSuntimes();
        }
        this.#addKeybinding();
        this.#changeTime(this.#computeTime());
        debug.message('Timer enabled.');
    }

    disable() {
        debug.message('Disabling Timer...');
        this.#removeKeybinding();
        this.#untrackSuntimes();
        this.#untrackLocation();
        this.#untrackTime();
        this.#disconnectSettings();
        this.#cancellable.cancel();
        debug.message('Timer disabled.');
    }


    get time() {
        return this.#time || Time.UNKNOWN;
    }

    #changeTime(time, manual = false) {
        if (time === this.#time) {
            if (!manual && this.#manuallySetTime)
                this.#manuallySetTime = false;
            return;
        }

        if (!manual && time !== this.#time && this.#manuallySetTime)
            return;

        this.#time = time;
        this.#manuallySetTime = manual;

        debug.message(manual ? `Time manually set to ${time}.` : `Time changed to ${time}.`);

        layoutManager.screenTransition.run();
        this.#interfaceSettings.set_string('color-scheme', time === Time.NIGHT ? 'prefer-dark' : 'default');
        this.notify('time');
    }


    #connectSettings() {
        debug.message('Connecting Timer to settings...');
        this.#settingsConnections.push({
            settings: this.#locationSettings,
            id: this.#locationSettings.connect('changed::enabled', this.#onLocationStateChanged.bind(this)),
        });
        this.#settingsConnections.push({
            settings: this.#settings,
            id: this.#settings.connect('changed::manual-schedule', this.#onManualScheduleStateChanged.bind(this)),
        });
        this.#settingsConnections.push({
            settings: this.#settings,
            id: this.#settings.connect('changed::nightthemeswitcher-ondemand-keybinding', this.#onOndemandKeybindingChanged.bind(this)),
        });
        this.#settingsConnections.push({
            settings: this.#interfaceSettings,
            id: this.#interfaceSettings.connect('changed::color-scheme', this.#onColorSchemeChanged.bind(this)),
        });
        // Only listen to the offset setting when not using a manual schedule
        if (!this.#settings.get_boolean('manual-schedule')) {
            this.#settingsConnections.push({
                settings: this.#settings,
                id: this.#settings.connect('changed::offset', this.#onOffsetChanged.bind(this)),
            });
        }
    }

    #disconnectSettings() {
        this.#settingsConnections.forEach(({ settings, id }) => settings.disconnect(id));
        this.#settingsConnections = [];
        debug.message('Disconnected Timer from settings.');
    }


    #trackTime() {
        debug.message('Watching for time change...');
        this.#timeTimeoutId = GLib.timeout_add_seconds(GLib.PRIORITY_DEFAULT, 1, () => {
            this.#changeTime(this.#computeTime());
            return GLib.SOURCE_CONTINUE;
        });
    }

    #untrackTime() {
        if (this.#timeTimeoutId) {
            GLib.Source.remove(this.#timeTimeoutId);
            this.#timeTimeoutId = null;
        }
        debug.message('Stopped watching for time change.');
    }


    #trackLocation() {
        debug.message('Connecting to GeoClue...');
        Geoclue.Simple.new(
            'org.gnome.Shell',
            Geoclue.AccuracyLevel.CITY,
            this.#cancellable,
            this.#onGeoclueReady.bind(this)
        );
    }

    #untrackLocation() {
        debug.message('Disconnecting from GeoClue...');
        if (this.#geoclue && this.#geoclueLocationConnectionId) {
            this.#geoclue.disconnect(this.#geoclueLocationConnectionId);
            this.#geoclueLocationConnectionId = null;
            this.#geoclue = null;
        }
        debug.message('Disconnected from GeoClue.');
    }


    #trackSuntimes() {
        debug.message('Regularly updating sun times...');
        this.#suntimesTimeoutId = GLib.timeout_add_seconds(GLib.PRIORITY_DEFAULT, 3600, () => {
            this.#updateSuntimes();
            return GLib.SOURCE_CONTINUE;
        });
    }

    #untrackSuntimes() {
        if (this.#suntimesTimeoutId) {
            GLib.Source.remove(this.#suntimesTimeoutId);
            this.#suntimesTimeoutId = null;
        }
        debug.message('Stopped regularly updating sun times.');
    }


    #addKeybinding() {
        this.#previousKeybinding = this.#settings.get_strv('nightthemeswitcher-ondemand-keybinding')[0];
        if (!this.#settings.get_strv('nightthemeswitcher-ondemand-keybinding')[0])
            return;
        debug.message('Adding keybinding...');
        wm.addKeybinding(
            'nightthemeswitcher-ondemand-keybinding',
            this.#settings,
            Meta.KeyBindingFlags.IGNORE_AUTOREPEAT,
            Shell.ActionMode.NORMAL | Shell.ActionMode.OVERVIEW,
            () => {
                const time = this.time === Time.NIGHT ? Time.DAY : Time.NIGHT;
                this.#changeTime(time, true);
            }
        );
        debug.message('Added keybinding.');
    }

    #removeKeybinding() {
        if (this.#previousKeybinding) {
            debug.message('Removing keybinding...');
            wm.removeKeybinding('nightthemeswitcher-ondemand-keybinding');
            debug.message('Removed keybinding.');
        }
    }


    #onLocationStateChanged() {
        this.disable();
        this.enable();
    }

    #onManualScheduleStateChanged() {
        this.disable();
        this.enable();
    }

    #onOffsetChanged() {
        this.#updateSuntimes();
    }

    #onOndemandKeybindingChanged() {
        this.#removeKeybinding();
        this.#addKeybinding();
    }

    #onColorSchemeChanged() {
        const time = this.#interfaceSettings.get_string('color-scheme') === 'prefer-dark' ? Time.NIGHT : Time.DAY;
        this.#changeTime(time, true);
    }

    #onGeoclueReady(_geoclue, result) {
        try {
            this.#geoclue = Geoclue.Simple.new_finish(result);
            this.#geoclueLocationConnectionId = this.#geoclue.connect('notify::location', this.#onLocationChanged.bind(this));
            debug.message('Connected to GeoClue.');
            this.#onLocationChanged();
        } catch (e) {
            const [latitude, longitude] = this.#settings.get_value('location').deepUnpack();
            if (latitude >= -90 && latitude <= 90 && longitude >= -180 && longitude <= 180) {
                console.error(`[${NTSMetadata.name}] Unable to retrieve the location, using the last known location instead.\n${e}`);
                this.#updateSuntimes();
            } else {
                console.error(`[${NTSMetadata.name}] Unable to retrieve the location, using the manual schedule times instead.\n${e}`);

                const source = new MessageTray.Source(NTSMetadata.name, 'dialog-information-symbolic');
                messageTray.add(source);

                const notification = new MessageTray.Notification(
                    source,
                    _('Unknown Location'),
                    _('A manual schedule will be used to switch the dark mode.'),
                    {
                        gicon: Gio.icon_new_for_string(GLib.build_filenamev([NTSMetadata.path, 'icons', 'nightthemeswitcher-symbolic.svg'])),
                    }
                );
                notification.addAction(_('Edit Manual Schedule'), () => this.#openPrefs());

                const locationSettingsApp = Shell.AppSystem.get_default().lookup_app('gnome-location-panel.desktop');
                if (locationSettingsApp)
                    notification.addAction(_('Open Location Settings'), () => locationSettingsApp.activate());

                notification.connect('activated', () => this.#openPrefs());

                source.showNotification(notification);

                this.#settings.set_boolean('manual-schedule', true);
            }
        }
    }

    #onLocationChanged(_geoclue, _location) {
        debug.message('Location has changed.');
        const { latitude, longitude } = this.#geoclue.get_location();
        this.#settings.set_value('location', new GLib.Variant('(dd)', [latitude, longitude]));
        debug.message(`Current location: (${latitude};${longitude})`);
        this.#updateSuntimes();
    }


    #computeTime() {
        const sunrise = this.#settings.get_double('sunrise');
        const sunset = this.#settings.get_double('sunset');
        const datetime = GLib.DateTime.new_now_local();
        const hour = datetime.get_hour() + datetime.get_minute() / 60 + datetime.get_second() / 3600;

        // Regular schedule
        if (sunrise < sunset)
            return hour >= sunrise && hour < sunset ? Time.DAY : Time.NIGHT;
        // Sunset happens on the day after
        else if (sunrise > sunset)
            return hour >= sunrise || hour < sunset ? Time.DAY : Time.NIGHT;
        else
            return Time.DAY;
    }

    #updateSuntimes() {
        const [latitude, longitude] = this.#settings.get_value('location').deepUnpack();

        if (latitude < -90 && latitude > 90 && longitude < -180 && longitude > 180)
            return;

        debug.message('Updating sun times...');

        const rad = degrees => degrees * Math.PI / 180;
        const deg = radians => radians * 180 / Math.PI;

        // Calculations from https://www.esrl.noaa.gov/gmd/grad/solcalc/calcdetails.html
        const dtNow = GLib.DateTime.new_now_local();
        const dtZero = GLib.DateTime.new_utc(1900, 1, 1, 0, 0, 0);

        const timeSpan = dtNow.difference(dtZero);

        const date = timeSpan / 1000 / 1000 / 60 / 60 / 24 + 2;
        const tzOffset = dtNow.get_utc_offset() / 1000 / 1000 / 60 / 60;

        const julianDay = date + 2415018.5 - tzOffset / 24;
        const julianCentury = (julianDay - 2451545) / 36525;
        const geomMeanLongSun = (280.46646 + julianCentury * (36000.76983 + julianCentury * 0.0003032)) % 360;
        const geomMeanAnomSun = 357.52911 + julianCentury * (35999.05029 - 0.0001537 * julianCentury);
        const eccentEarthOrbit = 0.016708634 - julianCentury * (0.000042037 + 0.0000001267 * julianCentury);
        const sunEqOfCtr = Math.sin(rad(geomMeanAnomSun)) * (1.914602 - julianCentury * (0.004817 + 0.000014 * julianCentury)) + Math.sin(rad(2 * geomMeanAnomSun)) * (0.019993 - 0.000101 * julianCentury) + Math.sin(rad(3 * geomMeanAnomSun)) * 0.000289;
        const sunTrueLong = geomMeanLongSun + sunEqOfCtr;
        const sunAppLong = sunTrueLong - 0.00569 - 0.00478 * Math.sin(rad(125.04 - 1934.136 * julianCentury));
        const meanObliqEcliptic = 23 + (26 + ((21.448 - julianCentury * (46.815 + julianCentury * (0.00059 - julianCentury * 0.001813)))) / 60) / 60;
        const obliqCorr = meanObliqEcliptic + 0.00256 * Math.cos(rad(125.04 - 1934.136 * julianCentury));
        const sunDeclin = deg(Math.asin(Math.sin(rad(obliqCorr)) * Math.sin(rad(sunAppLong))));
        const varY = Math.tan(rad(obliqCorr / 2)) * Math.tan(rad(obliqCorr / 2));
        const eqOfTime = 4 * deg(varY * Math.sin(2 * rad(geomMeanLongSun)) - 2 * eccentEarthOrbit * Math.sin(rad(geomMeanAnomSun)) + 4 * eccentEarthOrbit * varY * Math.sin(rad(geomMeanAnomSun)) * Math.cos(2 * rad(geomMeanLongSun)) - 0.5 * varY * varY * Math.sin(4 * rad(geomMeanLongSun)) - 1.25 * eccentEarthOrbit * eccentEarthOrbit * Math.sin(2 * rad(geomMeanAnomSun)));
        const haSunrise = deg(Math.acos(Math.cos(rad(90.833)) / (Math.cos(rad(latitude)) * Math.cos(rad(sunDeclin))) - Math.tan(rad(latitude)) * Math.tan(rad(sunDeclin))));
        const solarNoon = (720 - 4 * longitude - eqOfTime + tzOffset * 60) / 1440;

        const timeSunrise = solarNoon - haSunrise * 4 / 1440;
        const timeSunset = solarNoon + haSunrise * 4 / 1440;

        const modulo = (n, m) => ((n % m) + m) % m;

        const offset = this.#settings.get_double('offset');
        const sunrise = modulo(timeSunrise * 24 + offset, 24);
        const sunset = modulo(timeSunset * 24 - offset, 24);

        this.#settings.set_double('sunrise', sunrise);
        this.#settings.set_double('sunset', sunset);

        debug.message(`New sun times: (sunrise: ${sunrise}; sunset: ${sunset})`);
    }
}
